### PR TITLE
Do not assert

### PR DESCRIPTION
**Pull request by bsudekum**
Closes: https://github.com/mapbox/mapbox-navigation-ios/issues/1081

Although incrementing the step/leg index beyond the number of steps/legs denotes a serious issue, we should gently tell the developer something is wrong instead of asserting and crashing. 

/cc @hervouinc @mapbox/navigation-ios

### CODE REVIEW

**Comment by JThramer** [comment]
I'm not sure I agree with this change. I think a strong argument could be made that it's the developer's responsibility to do basic sanity checks before manipulating the model. If we've messaged you that "hey, the user arrived at waypoint X" (denoting that a leg has completed), it should be trivial to check "Hey, is this my final destination (last leg of route)? or an intermediate destination?"

An `assert()` seems pretty fair to me, as it will crash on debug (alerting developers to issues entirely within their domain of control), but will still fall through in production (where I suppose an argument could be made to add a `guard` _after_ the `assert` to avoid a crash in a release app)

But yeah, this seems like a slippery slope to me. If we start getting defensive about manipulating our model, where does that take us? do we put `guard` statements around every property in every model class we have?

I should call out that the narrative of the initial ticket is somewhat confusing. Did the user run into a bug? or do they just not like the pattern that they called out? If there's an issue in `RouteController.updateStepProgress(for:)`, or `RouteController.advanceStepIndex(to:)`, we should _definitely_ not merge this and fix the base issue instead.

We should discuss this as a team before merging. 

/cc @mapbox/navigation-ios @1ec5

**Comment by hervouinc** [comment]
Hi @JThramer.
Indeed, it crashes in release since the stepIndex is used to access the currentStep in the step array, and stepIndex is out of bounds.

The behaviour we have in our use case is:
- the user goes over the last waypoint of the route
- we choose to not reroute when asked by the routeController to its delegate, at least for a while (based on several local properties) since we know the user can stop after the waypoint
- the user's location is still handled by the routeController and leads to `updateRouteStepProgress` and then `advanceStepIndex`. 

In my opinion, `updateRouteStepProgress` or at least  `advanceStepIndex` should not be called in that situation.

**Comment by bsudekum** [comment]
@hervouinc this information is helpful. Let me see if i can reproduce the assertion locally.

File: MapboxCoreNavigation/RouteProgress.swift [hunk 1]
```diff
@@ -17,7 +17,10 @@ open class RouteProgress: NSObject {
      */
     @objc public var legIndex: Int {
         didSet {
-            assert(legIndex >= 0 && legIndex < route.legs.endIndex)
+            guard legIndex >= 0 && legIndex < route.legs.endIndex else {
+                print("legIndex - \(legIndex) out of range.")
```
Comment (1ec5) [inline MapboxCoreNavigation/RouteProgress.swift]: In addition to @JThramer’s comments, it should be emphasized that `print()` is no substitute for proper error handling. If not for a few existing occurrences that allow certain features to be optional, I’d be in favor of a linter rule against `print()`.

**Comment by bsudekum** [comment]
Going to go about this by fixing the root problem and keeping the assertion. Incrementing the leg/step index beyond the number of steps is incorrect and should not occur.

**Pull request closed without merging**